### PR TITLE
Move to zarr storage

### LINKED ISSUE #561: Update compression for current object store - move to Zarr
This might be low priority but if we have a large object store with the current NetCDF write system, could we include the functionality to read the current data, compress it and replace the existing NetCDFs with Zarr stores?

### PR DESCRIPTION
**Pull request by gareth-j**
* **Summary of changes** (Bug fix, feature, docs update, ...)

This will move the storage from NetCDFs on the file system to zarr stores.

* **Please check if the PR fulfills these requirements**

- [x] Closes #561 
- [x] [Tests added and passed](https://docs.openghg.org/development/python_devel.html#testing) if fixing a bug or adding a new feature
- [x] All code checks passing - `black --line-length 110` run over code, `mypy` and `flake8` not showing any errors
- [x] Documentation and tutorials updated/added
- [x] Added an entry in the latest `CHANGELOG.md` file if fixing a bug or adding a new feature

## TODO

> [!NOTE]
> This will be done in a separate PR

- [ ]  Add ability to elevate inlets as per the original `recombine_datasets` function, see `elevate_duplicate_attrs`, this 
can be moved closer to the user facing code where the `xr.Dataset` is created. - see #921 

Question: do we need this `elevate_inlet` functionality? Wasn't it just used for the ranking functionality?

- [x] Remove NetCDF comparison tests such as `test_datasource_compression`

### Files to concentrate on

As this PR covers a large swathe of the codebase the files to concentrate on are:

- `openghg.store.base._datasource.py` - a lot of the interaction with the new zarr store is done here
- `openghg.store.storage.*` - these have been added to handle the new zarr store, encoding, compression checks etc
- `openghg.store._footprints.py` - I've made some changes to enable opening of multiple files using `xr.open_mfdataset` but we need to think about it as it brings with it chunking issues which may not be so easy to smooth over / set good defaults for
- `openghg.dataobjects._basedata.py` - this now handles the slicing of data and the elevation of inlets etc as `recombine_datasets` did before

### Failing tests

> [!NOTE]
> These have all been resolved as of 2024-02-21

There are a few failing tests (notes below test).

```
tests/analyse/test_scenario.py::test_scenario_infer_inputs_ch4
 - AssertionError: assert False
 ```

This fails because I'm not sure how we want to handle flux data with a single timestamp now we do a time slice.
@rt17603 could this be somewhere the representative daterange function be used?

```
tests/store/test_datasource.py::test_surface_data_stored_and_dated_correctly
 - AssertionError: assert Timestamp('2020-12-01 22:31:30+0000', tz='UTC') == Timestamp('2020-12-01 22:32:29+0000', 
 tz='UTC')
```

Do we want to have some kind of tolerance checker? I think this error is due to the representative daterange function changing the end of the daterange to ensure things don't overlap.

```
tests/store/test_footprints.py::test_footprint_read_using_mfdataset
 - ValueError: invalid unit abbreviation: varies
 
```

Here we open two NetCDFs and they don't have continuous data so OpenGHG says the period == "varies" but we musn't have encountered this before as we get an pandas error.

```
tests/store/test_footprints.py::test_passing_chunks_reduces_memory_consumption
 - assert False
 
```

This is a placeholder for when we know what kind of chunking we're going to do.

### CODE REVIEW

**Comment by gareth-j** [comment]
TODO

- [ ]  Add ability to elevate inlets as per the original `recombine_datasets` function, see `elevate_duplicate_attrs`, this can be moved closer to the user facing code where the `xr.Dataset` is created.

**Comment by gareth-j** [comment]
I'll get these conflicts fixed.

**Review by rt17603 — COMMENTED** [review]
With this PR in particular, it would be good to have a summary of the changes and how this impacts the code base in the PR description. Particularly it would be great to make sure we understand what the aims and expectations around this code change are.

I've added a few small comments to address in this review. As mentioned above, if I can get a steer on what the overall changes are I can do a final quick check over before this is merged.

File: mypy.ini [hunk 1]
```diff
@@ -41,3 +41,9 @@ ignore_errors = True
 
 [mypy-rich]
 ignore_missing_imports = True
+
+[mypy-zarr]
+ignore_missing_imports = True
+
+[mypy-numcodecs]
+ignore_missing_imports = True
```
[the remark this answers is not in the document]
Comment (gareth-j) [reply mypy.ini:49]: Done

File: openghg/store/storage/_zarrstore.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        """Hash the data at the given key"""
+        pass
+
+    @abstractmethod
+    def get_hash(self, key: str) -> str:
+        """Get the hash of the data at the given key"""
+        pass
+
+    @abstractmethod
+    def hash_equal(self, key: str, dataset: Dataset) -> bool:
```
[the remark this answers is not in the document]
Comment (gareth-j) [reply openghg/store/storage/_zarrstore.py]: I've removed these functions as we don't currently have a wash of hashing Datasets reliably.

File: openghg/store/storage/_localzarrstore.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
+"""
+So this is a prototype for using zarr with the
```
[the remark this answers is not in the document]
Comment (gareth-j) [reply openghg/store/storage/_localzarrstore.py]: Done

File: openghg/store/storage/_zarrstore.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        pass
+
+    @abstractmethod
+    def store_key(self) -> str:
+        """Return the key of this zarr store"""
+        pass
+
+    @abstractmethod
+    def pop(self, key: str, version: str) -> Dataset:
+        """Pop some data from the store."""
```
[the remark this answers is not in the document]
Comment (gareth-j) [reply openghg/store/storage/_zarrstore.py]: Updated.

File: openghg/dataobjects/_basedata.py [hunk 5]
```diff
@@ -1,19 +1,87 @@
...
+        uuid: Optional[str] = None,
+        version: Optional[str] = None,
+        start_date: Optional[Union[str, Timestamp]] = None,
+        end_date: Optional[Union[str, Timestamp]] = None,
+        sort: bool = False,
+        elevate_inlet: bool = False,
+        attrs_to_check: Optional[Dict] = None,
+    ) -> None:
+        """
+        Args:
```
[the remark this answers is not in the document]
Comment (gareth-j) [reply openghg/dataobjects/_basedata.py:24]: Done.

File: tests/store/test_footprints.py [hunk 6]
```diff
@@ -213,13 +218,13 @@ def test_read_footprint_high_spatial_resolution():
         "min_longitude": -97.9,
         "max_latitude": 79.057,
         "min_latitude": 10.729,
-        "high_spatial_resolution": "True",
+        "high_spatial_resolution": True,
```
[the remark this answers is not in the document]
Comment (gareth-j) [reply tests/store/test_footprints.py:222]: Good catch, just fixed this. We'll store strings now. I've also added some write out to NetCDF sections to a couple of the tests so we make sure users can output things without issues.

**Comment by gareth-j** [comment]
For the broken tests

### test_surface_data_stored_and_dated_correctly

This is caused by the adding of the sampling period to the end of the last value by the representative daterange function

### test_footprint_read_using_mfdataset

We'll just need to add a bit of logic to fail gracefully here as we will want to store data that does vary.

**Review by SutarPrasad — CHANGES_REQUESTED** [review]
**The development achieves the goal of processing unseen files. However raises some new issues as summarised below**

**Summary**: On the new store when we `standardise_footprint` the list of filepaths with chunk value for time dimension it raises the `Value Error for inconsistent chunks` on other dimensions.
Further, if the chunks{"time":100, "lat":100, "lon":100} are passed for the first few files and then additional unprocessed file paths are passed to the same list of the file paths with only `{"time":100}` then it doesn't complain.

**Input function parameters passed to test:**
`standardise_footprint(site="MHD",
                      height="10m",
                      species="co2",
                      domain="EUROPE",
                      model="NAME",
                      filepath=ordered_files_list[8:10],
                      store="zarr_list_180",
                      continuous="False",
                      chunks={"time":100},
                      period="1H",
                      )`

1.	First-time standardising footprints raise below error:
`
INFO:openghg.store:Processing the following files:
[PosixPath('/group/chem/acrg/LPDM/fp_NAME/EUROPE/MHD-10magl_UKV_co2_EUROPE_201409.nc'), PosixPath('/group/chem/acrg/LPDM/fp_NAME/EUROPE/MHD-10magl_UKV_co2_EUROPE_201410.nc')]
WARNING:openghg.store:Opening a number of footprints as a single Dataset is currently an experimental feature and may result in chunking errors, slow operation or high memory usage.
INFO:openghg.store:Rechunking with chunks={'time': 100}
INFO:openghg.store:Updating high_time_resolution to True for co2 data
ERROR:openghg.store:<class 'ValueError'>, <traceback object at 0x7fbfa88735c0>
Traceback (most recent call last):
  File "/user/home/vq21425/openghg/footprints_100_list.py", line 13, in <module>
    standardise_footprint(site="MHD",
  File "/user/home/vq21425/openghg/openghg/standardise/_standardise.py", line 555, in standardise_footprint
    return standardise(
  File "/user/home/vq21425/openghg/openghg/standardise/_standardise.py", line 50, in standardise
    result = dc.read_file(filepath=filepath, **kwargs)
  File "/user/home/vq21425/openghg/openghg/store/_footprints.py", line 460, in read_file
    datasource_uuids = self.assign_data(
  File "/user/home/vq21425/openghg/openghg/store/base/_base.py", line 226, in assign_data
    datasource.add_data(
  File "/user/home/vq21425/openghg/openghg/store/base/_datasource.py", line 153, in add_data
    return self.add_timed_data(
  File "/user/home/vq21425/openghg/openghg/store/base/_datasource.py", line 228, in add_timed_data
    time_chunksize: Union[str, int] = data.chunksizes[time_coord][0]
  File "/user/home/vq21425/dev_openghg_env/lib/python3.10/site-packages/xarray/core/dataset.py", line 2615, in chunksizes
    return get_chunksizes(self.variables.values())
  File "/user/home/vq21425/dev_openghg_env/lib/python3.10/site-packages/xarray/core/common.py", line 2008, in get_chunksizes
    raise ValueError(
ValueError: Object has inconsistent chunks along dimension lat. This can be fixed by calling unify_chunks().`


2. I liked the idea of showing the processed and processing files in the terminal, however passing a large number of file paths the terminal gets crowded to show `processed files` and `processing files`, thus we need to have a better printing format. Maybe use `Rich` in this case and only show the filename instead of the whole PosixPath. This can be raised as a new issue if we think it's right to do so.

File: tests/retrieve/test_access.py [hunk 7]
```diff
@@ -387,7 +387,7 @@ def test_get_footprint(inlet_keyword, inlet_value):
     assert metadata["min_longitude"] == pytest.approx(float(footprint.lon.min()))
     assert metadata["max_latitude"] == pytest.approx(float(footprint.lat.max()))
     assert metadata["min_latitude"] == pytest.approx(float(footprint.lat.min()))
-    assert metadata["high_time_resolution"] == False
+    assert metadata["high_time_resolution"] == "false"
```
Comment (rt17603) [inline tests/retrieve/test_access.py:390]: What was the issue we had with metadata needing to be saved as strings? I feel like we did look into this but not sure how it got resolved.

If this is a requirement, for searching we'll need to be careful around this as would make data hard to search for using booleans. If not we should make sure this gets saved as a bool.

File: tests/store/storage/test_localzarrstore.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+
+        concat = xr.concat([ds, ds2], dim="time")
+        retrieved = store.get(version="v0").compute()
+
+        assert retrieved.equals(concat)
+        for var in retrieved.data_vars:
+            assert not np.sum(np.isnan(retrieved[var]))
+
+
+@pytest.mark.xfail(reason="Attributes of Datasets appended to zarr store may currently be overwritten")
```
Comment (rt17603) [inline tests/store/storage/test_localzarrstore.py:77]: Does this test need to be updated or is this functionality which should be checked? Can add a specific issue for this if needed.

File: tests/store/test_obssurface.py [hunk 9]
```diff
@@ -523,29 +463,18 @@ def test_read_noaa_metastorepack(bucket):
...
-        "1998-01-01-23:10:00+00:00_1998-12-31-19:50:00+00:00",
-        "1999-01-14-22:15:00+00:00_1999-12-31-23:35:00+00:00",
-        "2000-03-05-00:00:00+00:00_2000-11-04-22:30:00+00:00",
-        "2001-01-05-21:45:00+00:00_2001-12-06-12:00:00+00:00",
-        "2002-01-12-12:00:00+00:00_2002-01-12-12:00:01+00:00",  # Code added 1 second to ensure is not a range of 0 seconds.
-    ]
+    ch4_datasource = Datasource(bucket=bucket, uuid=uuid)
 
-    data = ch4_data["1998-01-01-23:10:00+00:00_1998-12-31-19:50:00+00:00"]
+    assert ch4_datasource.data_keys() == ["1993-06-17-00:12:30+00:00_2002-01-12-12:00:00+00:00"]
```
Comment (rt17603) [inline tests/store/test_obssurface.py:468]: What are the data keys being stored for this (and do they not include the representative date time anymore)?

**Review by rt17603 — CHANGES_REQUESTED** [review]
Looks good and almost ready to merge - a few things to address on this PR (e.g. merge conflict lines in documentation) and we can open issues for anything else which is outstanding.

File: tests/retrieve/test_access.py (hunk 7 above)
Comment (brendan-m-murphy) [reply tests/retrieve/test_access.py:390]: I think outside of the metastore, `search` and the `read_file` functions are calling `clean_string` on all metadata values, so booleans are transformed to lower case strings. If we didn't do this, I think `TinyDB` would store and retrieve booleans.

Comment (rt17603) [reply tests/retrieve/test_access.py:390]: Ok great and looks like we have a test to check this as well - https://github.com/openghg/openghg/blob/Iss561/addZarrStore/tests/retrieve/test_search.py#L343.

Something we may want to look at again in future but seems to work ok for now.

**Comment by rt17603** [comment]
After #917 is merged, would be good to focus on addressing the current review to work towards getting this into the `new_object_store` branch (PR #831 ) so any more PRs for v0.8.0 release can work off that branch.

File: doc/source/tutorials/local/Adding_new_data/Adding_ancillary_data.rst [hunk 10]
```diff
@@ -5,8 +5,12 @@ This tutorial demonstrates how to add ancillary spatial data to the
 OpenGHG store. These are split into several data types which currently
 include:
 
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (gareth-j) [reply doc/source/tutorials/local/Adding_new_data/Adding_ancillary_data.rst:8]: Fixed

File: openghg/dataobjects/_basedata.py [hunk 11]
```diff
@@ -1,19 +1,107 @@
...
+            raise ValueError("Must supply either data or uuid and version")
+
+        self.metadata = metadata
+        self._uuid = uuid
+
+        self._start_date = start_date
+        self._end_date = end_date
+        self._zarrstore = None
+
+        if elevate_inlet:
```
[the remark this answers is not in the document]
Comment (gareth-j) [reply openghg/dataobjects/_basedata.py:59]: Added issue: https://github.com/openghg/openghg/issues/921

[the remark this answers is not in the document]
Comment (gareth-j) [reply openghg/dataobjects/_basedata.py:59]: Opened an issue https://github.com/openghg/openghg/issues/922

File: openghg/__init__.py [hunk 12]
```diff
@@ -2,7 +2,8 @@
 import os as _os
 import sys as _sys
 from pathlib import Path as _Path
-from rich.logging import RichHandler as _RichHandler
+
+# from rich.logging import RichHandler as _RichHandler
```
[the remark this answers is not in the document]
Comment (gareth-j) [reply openghg/__init__.py]: Done

File: openghg/store/storage/_localzarrstore.py [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+            chunking = self.match_chunking(version=version, dataset=dataset)
+            if chunking:
+                dataset = dataset.chunk(chunking)
+
+            dataset.to_zarr(
+                store=self._stores[version], mode="a", consolidated=True, append_dim=append_dim, compute=True
+            )
+        # Otherwise we create a new zarr Store for the version
+        else:
+            if not self._stores and version != "v0":
```
[the remark this answers is not in the document]
Comment (gareth-j) [reply openghg/store/storage/_localzarrstore.py:167]: Opened an issue: https://github.com/openghg/openghg/issues/922

File: openghg/objectstore/_local_store.py [hunk 14]
```diff
@@ -32,6 +34,21 @@
 logger.setLevel(logging.DEBUG)  # Have to set level for logger as well as handler
 
 
+# # TODO - context manage this? As writes are atomic not sure it needs to be closed
```
[the remark this answers is not in the document]
Comment (gareth-j) [reply openghg/objectstore/_local_store.py:37]: Removed.

File: tests/store/storage/test_localzarrstore.py (hunk 8 above)
Comment (gareth-j) [reply tests/store/storage/test_localzarrstore.py:77]: Added an issue: #923

File: openghg/retrieve/_search.py [hunk 15]
```diff
@@ -578,30 +577,26 @@ def _base_search(**kwargs: Any) -> SearchResults:
...
         # Remove once more comprehensive tests are done
         dupe_uuids = [k for k in metadata if k in general_metadata]
         if dupe_uuids:
             raise ObjectStoreError("Duplicate UUIDs found between buckets.")
 
         general_metadata.update(metadata)
 
-    return SearchResults(keys=dict(data_keys), metadata=dict(general_metadata), start_result="data_type")
+    return SearchResults(
+        metadata=dict(general_metadata), start_result="data_type", start_date=start_date, end_date=end_date
```
Comment (brendan-m-murphy) [inline openghg/retrieve/_search.py:601]:
```suggestion
        metadata=general_metadata, start_result="data_type", start_date=start_date, end_date=end_date
```

`general_metadata` is declared as a dict on line 534, so we don't need to convert it?

File: tests/store/test_obssurface.py (hunk 9 above)
Comment (gareth-j) [reply tests/store/test_obssurface.py:468]: They're used to store the daterange of the dataset as it's added to the Datasource. The daterange is given by the `get_representative_daterange_str` function.

File: openghg/dataobjects/_basedata.py [hunk 16]
```diff
@@ -2,19 +2,106 @@
...
+        """
+        This handles data for each of the data type classes. It accepts either a Dataset
+        or a UUID and version to lazy load a Dataset from a zarr store. If dates are passed then the
+        dataset is sliced to the requested time period.
+
+        Args:
+            metadata: Dictionary of metadata
+            data: Dataset if data is already loaded
+            uuid: UUID of Datasource to retrieve data from
+            version: Version of data requested from Datasrouce
```
Comment (brendan-m-murphy) [inline openghg/dataobjects/_basedata.py:37]:
```suggestion
            version: Version of data requested from Datasource
```

File: openghg/store/base/_datasource.py [hunk 17]
```diff
@@ -27,31 +30,55 @@ class Datasource:
     _datasource_root = "datasource"
     _data_root = "data"
```
Comment (brendan-m-murphy) [inline openghg/store/base/_datasource.py:31]:
```suggestion
```
`_data_root` isn't used any more; this info has moved to `LocalZarrStore`

File: openghg/store/base/_datasource.py [hunk 18]
```diff
@@ -567,53 +628,9 @@ def load_dataset(bucket: str, key: str) -> xr.Dataset:
...
-        d._stored = False
-
-        d.update_daterange()
-
-        return d
+        raise NotImplementedError(
+            "Loading of data directly from Datasource no longer supported. Use memory_store to access data stored in zarr store."
+        )
 
     def define_version_key(self, version: str) -> str:
```
Comment (brendan-m-murphy) [inline openghg/store/base/_datasource.py:635]: `define_version_key`, `define_data_key`, and `define_backup_version` aren't used anywhere

File: openghg/store/base/_datasource.py [hunk 19]
```diff
@@ -705,48 +682,49 @@ def key(self) -> str:
...
+        if version == "latest":
+            version = self._latest_version
 
-        return cls.from_data(bucket=bucket, data=data, shallow=shallow)
+        return self._store.copy_to_memorystore(version=version)
 
-    def data(self) -> Dict:
-        """Get the data stored in this Datasource
+    def get_data(self, version: str = "latest") -> xr.Dataset:
+        """Directly accessing a Datasource's data is no longer supported.
```
Comment (brendan-m-murphy) [inline openghg/store/base/_datasource.py:707]: This docstring is confusing: it seems like `get_data` loads the latest version of the data using `xr.open_zarr` (via `LocalZarrStore`)

Comment (brendan-m-murphy) [reply openghg/store/base/_datasource.py:707]: Is the difference between `copy_to_memorystore` and `get_data` that the memory store should be used when modifying data? For computations, it seems like `get_data` is sufficient.

Should `copy_to_memorystore` be more "internal" to the storage parts of openghg, and `get_data` be the interface used by retrieve and analyse?

File: openghg/store/base/_datasource.py (hunk 17 above)
Comment (gareth-j) [reply openghg/store/base/_datasource.py:31]: Removed.

File: openghg/store/base/_datasource.py [hunk 20]
```diff
@@ -946,11 +825,11 @@ def keys_in_daterange_str(self, daterange: str) -> List[str]:
         start_date = timestamp_tzaware(split_daterange[0])
         end_date = timestamp_tzaware(split_daterange[1])
 
-        data_keys = self._data_keys["latest"]["keys"]
+        data_keys = self._data_keys[self._latest_version]
 
         return self.key_date_compare(keys=data_keys, start_date=start_date, end_date=end_date)
 
-    def key_date_compare(self, keys: Dict[str, str], start_date: Timestamp, end_date: Timestamp) -> List:
+    def key_date_compare(self, keys: List, start_date: Timestamp, end_date: Timestamp) -> List:
```
Comment (brendan-m-murphy) [inline openghg/store/base/_datasource.py:832]: I think `key_date_compare` is only used in `keys_in_daterange` and `keys_in_daterange_str`, and neither of those are used anywhere.

File: openghg/store/base/_datasource.py (hunk 19 above)
Comment (gareth-j) [reply openghg/store/base/_datasource.py:707]: Yeah, I've made `copy_to_memorystore` internal for now and we can see later how we want to use it. I've also updated those docstrings.

File: openghg/store/base/_datasource.py (hunk 20 above)
Comment (gareth-j) [reply openghg/store/base/_datasource.py:814]: OK I've opened an issue for this: https://github.com/openghg/openghg/issues/926

**Pull request merged**